File: lua/test/moon_entity_test.lua
-- Moon entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("solardemo_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("MoonEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:Moon(nil)
    assert.is_not_nil(ent)
  end)

  -- Feature #4: the entity stream(action, ...) method runs the op pipeline and
  -- returns an iterator over result items. With the streaming feature active it
  -- yields the feature's incremental output; otherwise it falls back to the
  -- materialised list so stream always yields.
  it("should stream", function()
    local seed = {
      entity = {
        ["moon"] = {
          s1 = { id = "s1" },
          s2 = { id = "s2" },
          s3 = { id = "s3" },
        },
      },
    }

    -- Fallback: streaming inactive -> yields the materialised list items.
    local base = sdk.test(seed, nil)
    local seen = {}
    for item in base:Moon(nil):stream("list", nil, nil) do
      table.insert(seen, item)
    end
    assert.are.equal(3, #seen)

    -- Inbound: streaming active -> yields each item from the feature.
    local config = require("config_shared")()
    if type(config.feature) == "table" and config.feature.streaming ~= nil then
      local streamsdk = sdk.test(seed, { feature = { streaming = { active = true } } })
      local got = {}
      for item in streamsdk:Moon(nil):stream("list", nil, nil) do
        if vs.islist(item) then
          for _, sub in ipairs(item) do
            table.insert(got, sub)
          end
        else
          table.insert(got, item)
        end
      end
      assert.are.equal(3, #got)
    end
  end)

  it("should run basic flow", function()
    local setup = moon_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "list", "update", "load", "remove"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "moon." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set SOLARDEMO_TEST_MOON_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local moon_ref01_ent = client:Moon(nil)
    local moon_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.moon"), "moon_ref01"))
    moon_ref01_data["planet_id"] = setup.idmap["planet01"]

    local moon_ref01_data_result, err = moon_ref01_ent:create(moon_ref01_data, nil)
    assert.is_nil(err)
    moon_ref01_data = helpers.to_map(type(moon_ref01_data_result) == 'table' and moon_ref01_data_result.data_get and moon_ref01_data_result:data_get() or moon_ref01_data_result)
    assert.is_not_nil(moon_ref01_data)
    assert.is_not_nil(moon_ref01_data["id"])

    -- LIST
    local moon_ref01_match = {
      ["planet_id"] = setup.idmap["planet01"],
    }

    local moon_ref01_list_result, err = moon_ref01_ent:list(moon_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(moon_ref01_list_result)

    local found_item = vs.select(
      runner.entity_list_to_data(moon_ref01_list_result),
      { id = moon_ref01_data["id"] })
    assert.is_false(vs.isempty(found_item))

    -- UPDATE
    local moon_ref01_data_up0_up = {
      id = moon_ref01_data["id"],
      ["planet_id"] = setup.idmap["planet_id"],
    }

    local moon_ref01_markdef_up0_name = "kind"
    local moon_ref01_markdef_up0_value = "Mark01-moon_ref01_" .. tostring(setup.now)
    moon_ref01_data_up0_up[moon_ref01_markdef_up0_name] = moon_ref01_markdef_up0_value

    local moon_ref01_resdata_up0_result, err = moon_ref01_ent:update(moon_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local moon_ref01_resdata_up0 = helpers.to_map(type(moon_ref01_resdata_up0_result) == 'table' and moon_ref01_resdata_up0_result.data_get and moon_ref01_resdata_up0_result:data_get() or moon_ref01_resdata_up0_result)
    assert.is_not_nil(moon_ref01_resdata_up0)
    assert.are.equal(moon_ref01_resdata_up0["id"], moon_ref01_data_up0_up["id"])
    assert.are.equal(moon_ref01_resdata_up0[moon_ref01_markdef_up0_name], moon_ref01_markdef_up0_value)

    -- LOAD
    local moon_ref01_match_dt0 = {
      id = moon_ref01_data["id"],
    }
    local moon_ref01_data_dt0_loaded, err = moon_ref01_ent:load(moon_ref01_match_dt0, nil)
    assert.is_nil(err)
    local moon_ref01_data_dt0_load_result = helpers.to_map(type(moon_ref01_data_dt0_loaded) == 'table' and moon_ref01_data_dt0_loaded.data_get and moon_ref01_data_dt0_loaded:data_get() or moon_ref01_data_dt0_loaded)
    assert.is_not_nil(moon_ref01_data_dt0_load_result)
    assert.are.equal(moon_ref01_data_dt0_load_result["id"], moon_ref01_data["id"])

    -- REMOVE
    local moon_ref01_match_rm0 = {
      id = moon_ref01_data["id"],
    }
    local _, err = moon_ref01_ent:remove(moon_ref01_match_rm0, nil)
    assert.is_nil(err)

    -- LIST
    local moon_ref01_match_rt0 = {
      ["planet_id"] = setup.idmap["planet01"],
    }

    local moon_ref01_list_rt0_result, err = moon_ref01_ent:list(moon_ref01_match_rt0, nil)
    assert.is_nil(err)
    assert.is_table(moon_ref01_list_rt0_result)

    local not_found_item = vs.select(
      runner.entity_list_to_data(moon_ref01_list_rt0_result),
      { id = moon_ref01_data["id"] })
    assert.is_true(vs.isempty(not_found_item))

  end)
end)

function moon_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/moon/MoonTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read moon test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "moon01", "moon02", "moon03", "planet01", "planet02", "planet03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("SOLARDEMO_TEST_MOON_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["SOLARDEMO_TEST_MOON_ENTID"] = idmap,
    ["SOLARDEMO_TEST_LIVE"] = "FALSE",
    ["SOLARDEMO_TEST_EXPLAIN"] = "FALSE",
  })

  local idmap_resolved = helpers.to_map(
    env["SOLARDEMO_TEST_MOON_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end
  if idmap_resolved["planet_id"] == nil then
    idmap_resolved["planet_id"] = idmap_resolved["planet01"]
  end

  if env["SOLARDEMO_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["SOLARDEMO_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["SOLARDEMO_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end

File: rb/test/moon_entity_test.rb
# Moon entity test

require "minitest/autorun"
require "json"
require_relative "../Solardemo_sdk"
require_relative "runner"

class MoonEntityTest < Minitest::Test
  def test_create_instance
    testsdk = SolardemoSDK.test(nil, nil)
    ent = testsdk.Moon(nil)
    assert !ent.nil?
  end

  # Feature #4: the entity stream(action, ...) method runs the op pipeline and
  # returns an Enumerator over result items. With the streaming feature active
  # it yields the feature's incremental output; otherwise it falls back to the
  # materialised list so stream always yields.
  def test_stream
    seed = {
      "entity" => {
        "moon" => {
          "s1" => { "id" => "s1" },
          "s2" => { "id" => "s2" },
          "s3" => { "id" => "s3" },
        },
      },
    }

    # Fallback: streaming inactive -> yields the materialised list items.
    base = SolardemoSDK.test(seed, nil)
    seen = base.Moon(nil).stream("list", nil, nil).to_a
    assert_equal 3, seen.length

    # Inbound: streaming active -> yields each item from the feature.
    cfg = SolardemoConfig.shared_config
    if cfg["feature"].is_a?(Hash) && cfg["feature"].key?("streaming")
      sdk = SolardemoSDK.test(seed, { "feature" => { "streaming" => { "active" => true } } })
      got = []
      sdk.Moon(nil).stream("list", nil, nil).each do |item|
        if item.is_a?(Array)
          got.concat(item)
        else
          got << item
        end
      end
      assert_equal 3, got.length
    end
  end

  def test_basic_flow
    setup = moon_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create", "list", "update", "load", "remove"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "moon." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set SOLARDEMO_TEST_MOON_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    moon_ref01_ent = client.Moon(nil)
    moon_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.moon"), "moon_ref01"))
    moon_ref01_data["planet_id"] = setup[:idmap]["planet01"]

    moon_ref01_data_result = moon_ref01_ent.create(moon_ref01_data, nil)
    moon_ref01_data = Helpers.to_map(moon_ref01_data_result.respond_to?(:data_get) ? moon_ref01_data_result.data_get : moon_ref01_data_result)
    assert !moon_ref01_data.nil?
    assert !moon_ref01_data["id"].nil?

    # LIST
    moon_ref01_match = {
      "planet_id" => setup[:idmap]["planet01"],
    }

    moon_ref01_list_result = moon_ref01_ent.list(moon_ref01_match, nil)
    assert moon_ref01_list_result.is_a?(Array)

    found_item = Vs.select(
      Runner.entity_list_to_data(moon_ref01_list_result),
      { "id" => moon_ref01_data["id"] })
    assert !Vs.isempty(found_item)

    # UPDATE
    moon_ref01_data_up0_up = {
      "id" => moon_ref01_data["id"],
      "planet_id" => setup[:idmap]["planet_id"],
    }

    moon_ref01_markdef_up0_name = "kind"
    moon_ref01_markdef_up0_value = "Mark01-moon_ref01_#{setup[:now]}"
    moon_ref01_data_up0_up[moon_ref01_markdef_up0_name] = moon_ref01_markdef_up0_value

    moon_ref01_resdata_up0_result = moon_ref01_ent.update(moon_ref01_data_up0_up, nil)
    moon_ref01_resdata_up0 = Helpers.to_map(moon_ref01_resdata_up0_result.respond_to?(:data_get) ? moon_ref01_resdata_up0_result.data_get : moon_ref01_resdata_up0_result)
    assert !moon_ref01_resdata_up0.nil?
    assert_equal moon_ref01_resdata_up0["id"], moon_ref01_data_up0_up["id"]
    assert_equal moon_ref01_resdata_up0[moon_ref01_markdef_up0_name], moon_ref01_markdef_up0_value

    # LOAD
    moon_ref01_match_dt0 = {
      "id" => moon_ref01_data["id"],
    }
    moon_ref01_data_dt0_loaded = moon_ref01_ent.load(moon_ref01_match_dt0, nil)
    moon_ref01_data_dt0_load_result = Helpers.to_map(moon_ref01_data_dt0_loaded.respond_to?(:data_get) ? moon_ref01_data_dt0_loaded.data_get : moon_ref01_data_dt0_loaded)
    assert !moon_ref01_data_dt0_load_result.nil?
    assert_equal moon_ref01_data_dt0_load_result["id"], moon_ref01_data["id"]

    # REMOVE
    moon_ref01_match_rm0 = {
      "id" => moon_ref01_data["id"],
    }
    moon_ref01_ent.remove(moon_ref01_match_rm0, nil)

    # LIST
    moon_ref01_match_rt0 = {
      "planet_id" => setup[:idmap]["planet01"],
    }

    moon_ref01_list_rt0_result = moon_ref01_ent.list(moon_ref01_match_rt0, nil)
    assert moon_ref01_list_rt0_result.is_a?(Array)

    not_found_item = Vs.select(
      Runner.entity_list_to_data(moon_ref01_list_rt0_result),
      { "id" => moon_ref01_data["id"] })
    assert Vs.isempty(not_found_item)

  end
end

def moon_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "moon", "MoonTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = SolardemoSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["moon01", "moon02", "moon03", "planet01", "planet02", "planet03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["SOLARDEMO_TEST_MOON_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "SOLARDEMO_TEST_MOON_ENTID" => idmap,
    "SOLARDEMO_TEST_LIVE" => "FALSE",
    "SOLARDEMO_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["SOLARDEMO_TEST_MOON_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end
  if idmap_resolved["planet_id"].nil?
    idmap_resolved["planet_id"] = idmap_resolved["planet01"]
  end

  if env["SOLARDEMO_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      # FIRST, so the generated fields below win: sdk-test-control.json's
      # test.client.options adds to the live client, it does not redirect it.
      Runner.live_client_options,
      {
      },
      extra || {},
    ])
    client = SolardemoSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["SOLARDEMO_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["SOLARDEMO_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
